### DISCUSSION TITLE
Graphql with axios

### DISCUSSION THREAD

**Discussion by BernardA** [mswjs/msw#879] [Q&A]
This is more of a question that an issue per se.

I have a Next js app that uses graphql to communicate with the API - both my own API and Wordpress'. 

In both cases no graphql client like Apollo is used. 

Instead I use `axios` to make the query or mutations, like in the instance below:

 At `/pages/brands`:

      import { apiQl } from 'lib/functions';
      .....

       const Brands = ({ brands }) => {

        .......
       };

         const queryQl = `query getBrandsModels {
                brands(
                      isActive: true
                     _order: {brand: "ASC"}
              ) {
                  id
                  brand
                  models(
                     isActive: true
                    _order: {model: "ASC"}
                 ){
                     id
                     model
             }
          }
        }`;

        export async function getServerSideProps() {
             const data = await apiQl(queryQl, null, false);
             return {
                  props: {
                        brands: data.data.brands,
                 },
              };
          }

Where `apiQl` is like: 

         export const apiQl = (data, variables = null, isForage = null) => {
               return axios
                    .post(process.env.NEXT_PUBLIC_API_GRAPHQL_URL, {
                        query: data,
                       variables,
                    })
                 .then((response) => {
                      return response.data;
                });
      };

This serves my purposes quite well. 

I have all the app state managed with `redux` and certainly see no reason to add a tool like `Apollo client`.

I have not been able to make `msw` work with my API setup and was wondering if it is possible at all without installing the grapqhql clients you prescribe which, as said, I have no reason to use otherwise.

If it is possible with my API setup, please indicated how.

Thanks

**Comment by kettanaito** [comment]
Hey, @BernardA.

The beauty of MSW is that it's agnostic of your request-issuing setup. All you have to do is to declare a query name that you wish to intercept and describe how to mock its response.

```js
import { graphql } from 'msw'

const handlers = [
  graphql.query('getBrandsModels', (req, res, ctx) => {
    return res(
      ctx.data({
        brands: [
          {
            id: 1,
            brand: 'Lego',
            models: [
              {
                id: 2,
                model: 'Porsche',
              },
            ],
          },
        ],
      })
    )
  }),
]
```

If you wish, you can scope the queries intercepted to a specific endpoint by using [graphql.link](https://mswjs.io/docs/api/graphql/link):

```js
import { graphql } from 'msw'

const myService = graphql.link(process.env.NEXT_PUBLIC_API_GRAPHQL_URL)

const handlers = [
  myService.query('getBrandsModels', (req, res, ctx) => {
    // The mocked response as described above.
  }),
]
```

**You then use the handlers above for both [browser and Node.js integrations](https://mswjs.io/docs/getting-started/integrate)**.

Let me know if you have any more questions or find this setup not working for you.

**Reply by kettanaito** [reply]
Also, make sure you're performing a correct GraphQL operation:

- Has `Content-Type: application/json`;
- Has the query string containing a `query` parameter (for GET request), or an object containing a `query` property (for POST request).
- Aforementioned `query` property contains a valid GraphQL schema.

**Reply by BernardA** [reply]
Thanks @kettanaito but it is not working.

It is not populating the `brands` property required by the page:



>  at Object.<anonymous> (tests/pages/brands.test.js:33:9)

    console.error
      Warning: Failed prop type: The prop `brands` is marked as required in `Brands`, but its value is `undefined`.
          at Brands (/Users/bernardo/Sites/soeezauto-admin/src/pages/brands/index.js:56:19)


As you can see below I tried to `console.log` the `req` to inspect it but nothing is output, so I was wondering if the `handlers` are being taken in by `server` at all. 

The test file for `/pages/brands` looks like below:

       import { graphql } from 'msw';
       import { setupServer } from 'msw/node';
       import { render, screen, waitFor } from '@testing-library/react';
       import '@testing-library/jest-dom';
       import Brands from 'pages/brands';

      const handlers = [
         graphql.query('getBrandsModels', (req, res, ctx) => {
             console.log('req', req); //  *** no output ***
             return res(
                 ctx.set('Content-Type', 'application/json'),
                 ctx.data({
                     brands: [
                         {
                             id: 1,
                             brand: 'Lego',
                             models: [
                                 {
                                     id: 2,
                                     model: 'Porsche',
                                 },
                             ],
                         },
                     ],
                  }),
              );
          }),
     ];

      const server = setupServer(...handlers);
      console.log('SERVER', server);

      beforeAll(() => server.listen());
     afterEach(() => server.resetHandlers());
     afterAll(() => server.close());

     describe('brands page', () => {
         it('should render table with contents', async () => {
             render(<Brands />);
             await waitFor(() => {
                 const main = screen.getByText('Alfa Romeo');
                 expect(main).toBeInTheDocument();
             });
         });
     });

The only global configuration I have setup for tests is:

 `/jest.config.js`

       module.exports = {
           testPathIgnorePatterns: ['<rootDir>/.next/', '<rootDir>/node_modules/'],
           moduleDirectories: ['node_modules', 'src'],
          roots: ['<rootDir>/src/', '<rootDir>/tests/'],
           testEnvironment: 'jsdom',
          moduleNameMapper: { '\\.(css|less)$': 'identity-obj-proxy' },
          testMatch: ['**/__tests__/**/*.[jt]s?(x)', '**/?(*.)+(test).[jt]s?(x)'],
      };

I have also run `npx msw init ./public`, which placed the SW on `/public`.

**Reply by kettanaito** [reply]
Are there any messages from MSW when running your tests? Can you please confirm that you're indeed matching that GraphQL query?

From the looks of it, your handler doesn't match the query.

**Reply by BernardA** [reply]
No messages from `msw`.

I tried both to `console.log`  the request variables and getting any `onUnhandledRequest` but no output in either case. See below:

        const handlers = [
                graphql.query('getBrandsModels', (req, res, ctx) => {
                 console.log('req', req.variables); *** NO OUTPUT**
                  .....

 And: 

         beforeAll(() => {
             server.listen({
                 onUnhandledRequest(req) {
                      console.error(
                          'Found an unhandled %s request to %s',
                            req.method,
                            req.url.href,
                     );
               },
           });
       });

Concerning the GrapqhQl query, note that my backend are both Symfony API-Platform and Wordpress GraphQl which certainly conform with `graphQl` specs.

The  `apiQl` function is the same regardless if it refers to a `query` or a `mutation` as the signature is identical for both. Also the method is always `POST`, of course.
In the function declaration below, `data` is the query string - not a particularly good parameter name, I know.

         export const apiQl = (data, variables = null, isForage = null) => {
           return axios
                .post(process.env.NEXT_PUBLIC_API_GRAPHQL_URL, {
                    query: data,
                   variables,
                })
             .then((response) => {
                  return response.data;
            });
        };

Looking at your docs, I understand that `msw` expects the following: 
    
         interface GraphQLRequest {
             query: string
            variables?: Record<string, any>
         }

So, while `query` is a string in both `msw` and `API-Platform`, `variables` seem to differ. 
In both cases `variables` is optional, but in the case of `API-Platform` - and Wordpress, it should be an object if it exists.
I am not quite sure of what `variables?: Record<string, any>` means in `msw` s case. 

It should not make a difference for this specific request though, as `variables` is `null`. 

Concerning the query name, it is the same, ie `getBrandsModels`.

**Reply by BernardA** [reply]
Just to mention that I added a `server.printHandlers();` with the following output:


          console.log
                 [graphql] query getBrandsModels (origin: *)
                  Declaration: /Users/bernardo/Sites/soeezauto-admin/tests/pages/brands.test.js:11:13

                 at node_modules/msw/node/lib/index.js:7912:29
               at Array.forEach (<anonymous>)

**Reply by kettanaito** [reply]
@BernardA, can you please put this into a reproduction repository? I'd be happy to look into it and help you figure out what's going on. Doing so will also help you discover issues, as isolation is one of the great debugging techniques.

**Reply by BernardA** [reply]
I appreciate you taking the time to assist on this.

I have created a repository with the code and it is available [here](https://github.com/BernardA/msw). 

After further reading your docs I could not find anything that would indicate where the problem might be. So I am at a bit of a loss here.

Should you need to consult the API schema, it is available publicly at www.soeezauto.com/api/graphql.

**Reply by kettanaito** [reply]
I'll take a look at the repository and let you know in this thread. Thanks for preparing it.

**Reply by kettanaito** [reply]
After looking into your repository I’ve found a couple of issues:

1. You weren’t using the latest version of the `msw` package. The latest one is `0.35.0`. Always run `npm i msw@latest` when in doubt.
2. The test run threw an error coming from your `<Brands/>` component:
```
Warning: Failed prop type: The prop `brands` is marked as required in `Brands`, but its value is `undefined`.
        at Brands (/Users/kettanaito/Projects/contrib/msw-879-graphql-match/src/pages/brands/index.js:56:19)
```
As the warning suggests, you weren’t passing the required `brands` prop when rendering the `<Brands/>` component in your test. I suspect anything can go wrong if you don’t provide components with the props they expect. I’ve removed `.isRequired` from its prop type and continued.
3. You make the GraphQL request in the `getServerSideProps` hook of NextJS. That hook is _not_ executed when you’re rendering `<Brands/>` in isolation (like you should do in the integration test).

If you wish to test the logic in the `getServerSideProps` hook, you need to do so in an end-to-end test. This implies running your page (not a single component) in a browser, which will execute all the server-side hooks of NextJS accordingly.

> NextJS is currently working on the official testing guidelines, so you may be able to test components with server-side hooks an in integration test. That is not possible now, however. Keep an eye on the NextJS announcements regarding this.

Take a look at the [NextJS usage example](https://github.com/vercel/next.js/tree/canary/examples/with-msw) to see how to set up MSW so that it’d intercept requests in server-side hooks like `getServerSideProps`. This setup needs to be done on the NextJS side, not in your tests. You should still use MSW as you do in your integration tests, just keep in mind what scope of the components you’re testing.

That being said, there’s a current limitation when it comes to controlling mocks on the server-side. NextJS runs a server in a separate process, so you cannot change the mocks you’ve declared on test runtime. This may not be an issue for you, but it’s worth mentioning. We’re pursuing the effort to create a recommended approach in #799 but it’s still in progress.

**Reply by BernardA** [reply]
I was quite aware I was not passing the props. I was assuming `msw` would intercept the request on `getServerSideProps` and pass it on. If you look at one of my comments above I had shown the error of the missing `brands` prop.

That what is suggested by [this Next js official example ](https://github.com/vercel/next.js/blob/canary/examples/with-msw/pages/index.js)as well.

That example clearly states that ` // Server-side requests are mocked by `mocks/server.js` which I have replicated on my code

If `msw` cannot handle that, than I certainly misunderstood its utility in this case and overall when testing Next js. As a rule I will always make all request necessary to render pages and their initial components  on `getServerSideProps` or 'getStaticProps`, ie, at page level. 

Doing otherwise would be wasting one of the big benefits of a tool like Next js.

Only when responding to an user action would a request originate at component level. 

But then, as you suggested, I would be better off testing end to end with a tool like `Cypress`.

Thanks again for your assistance

**Reply by kettanaito** [reply]
> I was assuming msw would intercept the request on `getServerSideProps` and pass it on.

MSW can not do that because `getServerSideProps` is not called when you test your component like that. It's called internally by NextJS that runs your app, which doesn't happen when testing components in isolation. Emphasizing that this has nothing to do with MSW. **You're trying to test something that's not executed**.

Once again, **server-side requests are mocked and function properly** in the NextJS example. You need to understand that statement relates to _End-to-End_ execution of your application. You're not running the entire NextJS app in your test.

I think you misunderstood my explanation completely. Your issue has nothing to do with MSW or its ability to intercept requests. When you render your component in isolation in a test like this:

```js
render(<Component/>)
```

You're extracting this component, isolating it, and effectively dropping any logic that lives outside of its scope (like `getServerSideProps`). **The issue is you trying to test server-side interaction that's invoked by NextJS server in an integration test where there's no such server**. For the lack of a better approach from NextJS team at the moment, you need to run your component on a page in a real browser to have methods like `getServerSideProps` be called. 

When you do run your page in a browser, you can use MSW to intercept _any_ requests, be it client-side or server-side, as long as you have it set up as shown in the NextJS usage example.
